### PR TITLE
move schema implementations to `core/schema`

### PR DESCRIPTION
**Pull request by vito**
`core/` is now for domain models only, reducing its dependency footprint and preventing import cycles.

The original motivation of this was to switch `project/*` from `*filesystem.Filesystem` to `*core.Directory`, but that introduced a cycle. We might not _need_ to do this if we just move most of `project/*` to core, but it feels more natural to move all the schema implementation code into `core/schema/` anyway.

This refactor is 95% just shifting code around, the only other change is that I added `(*core.Container).WithSecretVariable`, since that was implemented inline in the schema and relied on `(core.ContainerID).decode` which the schema no longer has access to.

### CODE REVIEW

File: core/container.go [hunk 1]
```diff
@@ -553,7 +550,7 @@ func (container *Container) ExitCode(ctx context.Context, gw bkgw.Client) (*int,
 		return nil, err
 	}
 
-	if file != nil {
+	if file == nil {
```
Comment (dolanor) [inline core/container.go:553]: if the file is `nil` doesn't it mean we should return some error?

**Review by dolanor — COMMENTED** [review]
Except for a small missunderstanding LGTM

File: core/container.go (hunk 1 above)
Comment (vito) [reply core/container.go:553]: This is part of the "happy path" from an API perspective: if the file doesn't exist, it just means the command hasn't run yet. The API returns a null `exitCode` in that case.

**Review by sipsma — APPROVED** [review]
LGTM

**Pull request merged**